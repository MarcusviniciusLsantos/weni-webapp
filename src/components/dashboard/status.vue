<template>
  <div class="weni-status">
    <UnnnicCard
      v-for="status in list"
      :key="`${status.service__type_service}:${status.id}`"
      :title="$t(`home.status.${status.service__type_service}`)"
      type="status"
      :scheme="statusSchemes[status.service__status.status]"
      :icon="statusIcons[status.service__type_service]"
      :description="
        status.service__status.intercurrence
          ? $t(`home.status.updated.${status.service__status.status}`, {
              time: timeAgo(status.service__status.intercurrence),
            })
          : $t('home.status.no_intercurrences_yet')
      "
      :status="$t(`home.status.title.${status.service__status.status}`)"
    />
  </div>
</template>

<script>
import { unnnicCallAlert } from '@weni/unnnic-system';
import { getTimeAgo } from '../../utils/plugins/timeAgo';
import { mapActions, mapGetters, mapState } from 'vuex';

export default {
  // eslint-disable-next-line vue/multi-word-component-names
  name: 'Status',
  data() {
    return {
      statusList: [],
      statusIcons: {
        type_service_chat: 'forum',
        type_service_inteligence: 'neurology',
        type_service_studio: 'ad',
        type_service_flows: 'account_tree',
      },
      loading: false,
      statusSchemes: {
        online: 'feedback-green',
        offline: 'feedback-red',
        intermittent: 'feedback-yellow',
        maintenance: 'feedback-blue',
      },
    };
  },
  mounted() {
    this.fetchStatus();
  },
  watch: {
    loading() {
      this.$emit('loadingStatus', this.loading);
    },
  },
  methods: {
    ...mapActions(['getStatus']),
    async fetchStatus() {
      try {
        this.loading = true;
        const response = await this.getStatus({
          projectUuid: this.currentProject.uuid,
        });
        this.statusList = response.data.results;
      } catch (e) {
        unnnicCallAlert({
          props: {
            text: this.$t('home.status_error'),
            title: 'Error',
            icon: 'check_circle',
            scheme: 'feedback-red',
            position: 'bottom-right',
            closeText: this.$t('close'),
          },
          seconds: 3,
        });
      } finally {
        this.loading = false;
      }
    },
    timeAgo(time) {
      const date = new Date(time);
      return getTimeAgo(date, this.profile.language);
    },
  },
  computed: {
    ...mapState({
      profile: (state) => state.Account.profile,
    }),

    ...mapGetters(['currentProject']),

    list() {
      const list = [];

      // register known services (also its ordering)
      const services = {
        type_service_flows: 'type_service_flows',
        type_service_studio: 'type_service_flows',
        type_service_inteligence: 'type_service_inteligence',
        type_service_chat: 'type_service_chat',
      };

      Object.keys(services).forEach((service) => {
        const serviceFound = this.statusList.find(
          (item) => item.service__type_service === services[service],
        );

        if (serviceFound) {
          list.push({
            ...serviceFound,
            service__type_service: service,
          });
        }
      });

      // if there is a service status that is unknown, add to list anyway
      this.statusList
        .filter((item) => {
          return !list.find(
            (service) =>
              service.service__type_service === item.service__type_service,
          );
        })
        .forEach((item) => list.push(item));

      return list;
    },
  },
};
</script>

<style lang="scss">
.weni-status {
  > * {
    background-color: $unnnic-color-neutral-lightest !important;
  }
  > :not(:last-child) {
    margin-bottom: $unnnic-spacing-stack-sm;
  }
}
</style>
