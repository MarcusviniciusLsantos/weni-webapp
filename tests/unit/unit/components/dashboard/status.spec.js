import { vi } from 'vitest';
import { shallowMount, createLocalVue, RouterLinkStub } from '@vue/test-utils';
import Vuex from 'vuex';

import { unnnicCallAlert as mockunnnicCallAlert } from '@weni/unnnic-system';

import status from '@/components/dashboard/status.vue';
import i18n from '@/utils/plugins/i18n';

import { project, profile } from '../../../__mocks__';

vi.mock('@weni/unnnic-system', () => ({
  unnnicCallAlert: vi.fn(),
}));

const localVue = createLocalVue();
localVue.use(Vuex);

describe('status.vue', () => {
  let wrapper;
  let store;
  let getters;
  let actions;
  let state;

  beforeEach(() => {
    getters = {
      currentProject() {
        return project;
      },
    };
    actions = {
      getStatus: vi.fn(),
    };
    state = {
      Account: {
        profile,
      },
    };
    store = new Vuex.Store({
      getters,
      actions,
      state,
    });
    wrapper = shallowMount(status, {
      localVue,
      store,
      i18n,
      mocks: {
        $t: () => 'some specific text',
        setTimeout: function () {
          return 99;
        },
        clearTimeout: function () {},
      },
      stubs: {
        RouterLink: RouterLinkStub,
        UnnnicCard: true,
      },
    });
  });

  it('renders a snapshot', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  describe('fetchStatus()', () => {
    it('test default value', async () => {
      const DATA = [
        {
          id: 481,
          project: 'a07f30ac-eab5-4ae6-9bb6-8f82c6c78087',
          service__status: {
            status: 'online',
            intercurrence: '2022-03-02T00:33:10.146802Z',
          },
          service__url: 'https://api.bothub.it/',
          service__default: true,
          service__last_updated: '2022-03-02T19:39:00.288738Z',
          service__type_service: 'type_service_inteligence',
          created_at: '2021-10-25T20:43:03.868486Z',
        },
      ];
      actions.getStatus.mockImplementation(() => {
        return {
          data: {
            results: DATA,
          },
        };
      });
      await wrapper.vm.fetchStatus();
      expect(wrapper.vm.statusList).toEqual(DATA);
    });

    it('should call unnnicCallAlert on error', async () => {
      actions.getStatus.mockImplementation(() => {
        throw new Error('error fetching');
      });
      expect(mockunnnicCallAlert).toHaveBeenCalled();
    });
  });

  // it('test computed projectUuid()', () => {
  //   const res = wrapper.vm.projectUuid;
  //   expect(res).toEqual(project.uuid);
  // });

  // it('test computed hasNext()', async () => {
  //   wrapper.setData({
  //     current: 2,
  //     total: 4,
  //   });
  //   const res = await wrapper.vm.hasNext;
  //   expect(res).toBeTruthy;
  // });

  // it('test computed hasPrevious()', async () => {
  //   wrapper.setData({
  //     current: 2,
  //   });
  //   const res = await wrapper.vm.hasPrevious;
  //   expect(res).toBeTruthy;
  // });

  // it('test method startAnimating()', async () => {
  //   await wrapper.vm.startAnimating();
  //   expect(wrapper.vm.animating).toBeTruthy;
  // });

  // it('test method onSelect()', async () => {
  //   const INDEX = 2;
  //   const spy = vi.spyOn(wrapper.vm, 'resetTimeout');
  //   await wrapper.vm.onSelect(INDEX);

  //   expect(spy).toHaveBeenCalledTimes(1);
  //   expect(wrapper.vm.current).toEqual(INDEX - 1);
  // });

  // it('test method onFinishAnimating()', async () => {
  //   const spy = vi.spyOn(wrapper.vm, 'next');
  //   await wrapper.vm.onFinishAnimating();

  //   expect(spy).toHaveBeenCalledTimes(1);
  //   expect(wrapper.vm.animating).toBeFalsy;
  // });
});
