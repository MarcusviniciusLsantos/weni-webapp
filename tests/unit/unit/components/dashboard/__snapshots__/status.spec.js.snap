// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`status.vue > renders a snapshot 1`] = `
<div
  class="weni-status"
>
  
  
</div>
`;
